warehouse/macaroons/services.py

```diff
@@ -9,20 +9,83 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from __future__ import annotations
 
+import base64
 import datetime
 import uuid
 
+from typing import TYPE_CHECKING
+
 import pymacaroons
 
 from pymacaroons.exceptions import MacaroonDeserializationException
 from sqlalchemy.orm import joinedload
 from zope.interface import implementer
 
+if TYPE_CHECKING:
+    from pyramid.request import Request
+
 from warehouse.macaroons import caveats
 from warehouse.macaroons.errors import InvalidMacaroonError
 from warehouse.macaroons.interfaces import IMacaroonService
 from warehouse.macaroons.models import Macaroon
+from warehouse.metrics import IMetricsService
+
+
+def _extract_basic_macaroon(auth):
+    """
+    A helper function for extracting a macaroon from a
+    HTTP Basic Authentication-style header.
+
+    Returns None if the header doesn't contain a structurally
+    valid macaroon, or the candidate (not yet verified) macaroon
+    in a serialized form.
+    """
+    try:
+        authorization = base64.b64decode(auth).decode()
+        auth_method, _, auth = authorization.partition(":")
+    except ValueError:
+        return None
+
+    if auth_method != "__token__":
+        return None
+
+    # Strip leading/trailing whitespace characters from the macaroon
+    auth = auth.strip()
+
+    return auth
+
+
+def _extract_http_macaroon(request: Request, increase_metrics: bool = False):
+    """
+    A helper function for the extraction of HTTP Macaroon from a given request.
+    Returns either a None if no macaroon could be found, or the string
+    that represents our serialized macaroon.
+    """
+    authorization = request.headers.get("Authorization")
+    if not authorization:
+        return None
+
+    try:
+        auth_method, auth = authorization.split(" ", 1)
+    except ValueError:
+        return None
+
+    auth_method = auth_method.lower()
+
+    if increase_metrics:
+        metrics = request.find_service(IMetricsService, context=None)
+        metrics.increment(
+            "warehouse.macaroon.auth_method", tags=[f"method:{auth_method}"]
+        )
+
+    if auth_method == "basic":
+        return _extract_basic_macaroon(auth)
+    elif auth_method in ["token", "bearer"]:
+        return auth
+
+    return None
```

Comment:
This LGTM behaviorally, but I think the `increase_metrics` flag is a bit of a wart 🙂 

Rather than internalizing metrics here, maybe we can increment this counter into an inline operation in the `identity(...)` method? That way we explicitly only ever call it once per authentication, without having to rely on a flag to do it.

Reply:
The problem is that `identity()` would need to know how to extract the authentication method from the request (in order to know which metric to increase). So the question is, do we want to extract that logic from `_extract_http_macaroon()`, and make it available to `identity()` ?

Reply:
Ah, right. Gross 🙂 

Hmm, in that case I suppose it's a bit cleaner to keep it behind that flag. I'll think about this some more.